lib/src/repository_url.dart

```diff
@@ -0,0 +1,180 @@
+  /// or throws [FormatException] if it was unable to recognize the pattern.
+  static RepositoryUrl parse(String input) {
+    final v = tryParse(input);
+    if (v == null) {
+      throw FormatException('Unable to parse repository url: `$input`.');
+    } else {
+      return v;
+    }
+  }
+
+  /// Resolves a relative path and returns a new instance of the [RepositoryUrl].
+  RepositoryUrl resolve(
+    String relativePath, {
+    String? separator,
+    String? branch,
+  }) {
+    final newPath = p.joinAll([
+      if (path.isNotEmpty) path,
+      if (relativePath.isNotEmpty) relativePath,
+    ]);
+    final normalizedNewPath = p.normalize(newPath);
+
+    if (RepositoryProvider.isGitHubCompatible(provider)) {
+      final extension = p.extension(normalizedNewPath).toLowerCase();
+      final needsRaw = _imageExtensions.contains(extension);
+      final newBaseUrl = baseUrl.endsWith('.git')
+          ? baseUrl.substring(0, baseUrl.length - 4)
+          : baseUrl;
+      return RepositoryUrl(
+        provider: provider,
+        baseUrl: newBaseUrl,
+        separator: separator ?? (needsRaw ? 'raw' : null) ?? 'blob',
+        branch: branch ?? this.branch ?? 'master',
+        path: normalizedNewPath,
+      );
+    } else {
+      return RepositoryUrl(
+        provider: provider,
+        baseUrl: baseUrl,
+        separator: separator,
+        branch: branch,
+        path: normalizedNewPath,
+      );
+    }
+  }
+
+  /// Creates a String representation of the URL.
+  String toUrl() =>
+      p.joinAll([baseUrl, separator, branch, path].whereType<String>());
+}
+
+abstract class RepositoryProvider {
+  static const github = 'github';
+  static const gitlab = 'gitlab';
+  static const unknown = 'unknown';
+
+  static bool isGitHubCompatible(String provider) =>
+      provider == github || provider == gitlab;
+}
+
+const _repoReplacePrefixes = {
+  'http://github.com': 'https://github.com',
+  'https://www.github.com': 'https://github.com',
+  'https://www.gitlab.com': 'https://gitlab.com',
+};
+
+const _githubSegmentSeparators = ['tree', 'blob', 'raw'];
+
+final _imageExtensions = <String>{'.gif', '.jpg', '.jpeg', '.png'};
+
+RepositoryUrl? _tryParseRepositoryUrl(String input) {
+  var url = input;
+
+  // apply known prefix replace patterns
+  for (final key in _repoReplacePrefixes.keys) {
+    if (url.startsWith(key)) {
+      url = url.replaceFirst(key, _repoReplacePrefixes[key]!);
+    }
+  }
+
+  final uri = Uri.parse(url);
+  final provider = _detectProvider(uri);
+
+  // detect repo vs path segments
+  final segments = Uri.parse(p.normalize(uri.path))
+      .pathSegments
+      .where((s) => s.isNotEmpty)
+      .toList();
+  final repoSegmentIndex = _repoSegmentIndex(provider, segments);
+  if (repoSegmentIndex < 0) return null;
+
+  String? separator;
+  final extraSegments = segments.skip(repoSegmentIndex + 1).toList();
+  if (extraSegments.isNotEmpty &&
+      _githubSegmentSeparators.contains(extraSegments.first)) {
```

Comment:
Should this also be conditioned on the provider being github?

Reply:
Done.